### PR TITLE
Upgrade to Pydantic 2

### LINKED ISSUE #273: Update to Pydantic2
Requires Dodal to have removed their pinning to v1.

### PR DESCRIPTION
**Pull request by callumforrester**
Fixes #273 

Supersedes #449 

Currently scanspec is spoofed, we need https://github.com/bluesky/scanspec/issues/122 merged before we can revert.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -17,7 +17,9 @@ dependencies = [
     "nslsii",
     "pyepics",
     "aioca",
-    "pydantic<2.0",        # Leave pinned until can check incompatibility
+    "pydantic>=2.0",
+    "scanspec>=0.7.0",
```
Comment (DiamondJoseph) [inline pyproject.toml:21]: Since we know scanspec 0.7.0 isn't compatible with serialising BaseModels like we need to, let's keep it at >=0.7.1

**Review by DiamondJoseph — CHANGES_REQUESTED** [review]

File: src/blueapi/service/main.py [hunk 2]
```diff
@@ -70,7 +70,6 @@ async def lifespan(app: FastAPI):
 
 app = FastAPI(
     docs_url="/docs",
-    on_shutdown=[teardown_runner],
```
[the remark this answers is not in the document]
Comment (ZohebShaikh) [reply src/blueapi/service/main.py:73]: https://fastapi.tiangolo.com/advanced/events/#shutdown-event

As we are using lifespan which declares both the startup and the shutdown ...
We cannot use both of them together

File: src/blueapi/cli/format.py [hunk 3]
```diff
@@ -94,6 +94,13 @@ def _describe_type(spec: dict[Any, Any], required: bool = False):
             if all_of := spec.get("allOf"):
                 items = (_describe_type(f, False) for f in all_of)
                 disp += f'{" & ".join(items)}'
+            elif any_of := spec.get("anyOf"):
```
[the remark this answers is not in the document]
Comment (ZohebShaikh) [reply src/blueapi/cli/format.py:97]: It was Callum's work ..should I change something here ?

[the remark this answers is not in the document]
Comment (DiamondJoseph) [reply src/blueapi/cli/format.py:97]: Just wanted to know the motivation. @callumforrester are we still reflecting optional fields such that it's clear they're optional? I know the handling for nulls were changed in fastapi

[the remark this answers is not in the document]
Comment (callumforrester) [reply src/blueapi/cli/format.py:97]: The handling of null unions was changed in pydantic2, see [this comment](https://github.com/pydantic/pydantic/issues/7161#issuecomment-2275671152) (and the issue in general).

File: src/blueapi/utils/base_model.py [hunk 4]
```diff
@@ -1,30 +1,19 @@
-from pydantic import BaseConfig, BaseModel, Extra
+from pydantic import BaseModel, ConfigDict
 
+# Pydantic config for blueapi API models with common config.
+BlueapiModelConfig = ConfigDict(
+    extra="forbid",
+    populate_by_name=True,
```
[the remark this answers is not in the document]
Comment (callumforrester) [reply src/blueapi/utils/base_model.py:6]: It was added to stop underscore fields being serialized. This appears to be the default behaviour in pydantic 2 now...

```python
>>> from pydantic import BaseModel
>>> class Foo(BaseModel):
...     bar: int
...     _baz: int
... 
>>> foo = Foo(bar=2, _baz=3)
>>> foo.model_dump()
{'bar': 2}

```

File: dev-requirements.txt [hunk 5]
```diff
@@ -103,19 +108,19 @@ mongoquery==1.4.2
...
 nslsii==0.10.3
-numcodecs==0.12.1
-numpy==1.26.4
+numcodecs==0.13.0
+numpy==2.0.1
 opencv-python-headless==4.10.0.84
 ophyd==1.9.0
-ophyd-async==0.3.4
-orjson==3.10.6
+ophyd-async==0.3.1
```
Comment (DiamondJoseph) [inline dev-requirements.txt:122]: also here. Would be nice to see those projects updated first

**Review by DiamondJoseph — APPROVED** [review]

**Pull request merged**